lib/pacemaker/pcmk_sched_instances.c

```diff
@@ -625,7 +595,116 @@ assign_instance(pe_resource_t *instance, const pe_node_t *prefer,
+        pe_rsc_debug(rsc, "Not assigning %s to current node %s: unavailable",
+                     instance->id, pe__node_name(current));
+        return false;
+    }
+
+    pe_rsc_trace(rsc, "Trying to assign %s to its current node %s",
+                 instance->id, pe__node_name(current));
+
+    /* The basic approach is to make a backup copy of instance->allowed_nodes,
+     * and then use instance->allowed_nodes as a scratch table while we try to
+     * assign instance to its current node.
+     *
+     * On each iteration, if instance gets assigned to a different node, we
+     * restore instance->allowed_nodes to its state from the beginning of the
+     * iteration. We reserve an instance for the assigned node, since assigning
+     * an instance to that node is more important. We then mark the assigned
+     * node as unavailable and repeat the process until we either succeed or run
+     * out of available instances.
+     */
+    allowed_orig = pcmk__copy_node_table(instance->allowed_nodes);
+
+    while (reserved < available) {
+        GHashTable *allowed_backup = NULL;  // Per-iteration backup
+        pe_node_t *chosen = NULL;
+
+        allowed_backup = pcmk__copy_node_table(instance->allowed_nodes);
+        chosen = assign_instance(rsc, instance, current, max_per_node);
+
+        if (pe__same_node(chosen, current)) {   // Success
+            g_hash_table_destroy(allowed_backup);
+            g_hash_table_destroy(allowed_orig);
+            return true;
+        }
+
+        /* Assignment failed, so give up. This might be impossible, as we might
+         * catch all such situations before this point.
+         */
+        if (chosen == NULL) {
+            pe_rsc_debug(rsc,
+                         "Not assigning %s to current node %s: unavailable",
+                         instance->id, pe__node_name(current));
+            g_hash_table_destroy(allowed_backup);
+            break;
+        }
+
+        // We prefer more strongly to assign an instance to some other node
+        pe_rsc_debug(rsc, "Not assigning %s to current node %s: %s is better",
+                     instance->id, pe__node_name(current),
+                     pe__node_name(chosen));
+
+        // Reserve one instance for the chosen node and try again
+        reserved++;
+        if (reserved >= available) {
+            pe_rsc_debug(rsc,
+                         "Not assigning %s to current node %s: "
+                         "other assignments took higher priority",
+                         instance->id, pe__node_name(current));
+        } else {
+            pe_rsc_debug(rsc,
+                         "Reserved an instance of %s for %s. Retrying "
+                         "assignment of %s to %s",
+                         rsc->id, pe__node_name(chosen), instance->id,
+                         pe__node_name(current));
+
+            // Make the chosen node unavailable for the next assignment attempt
```

Comment:
In the end, I realized I need to revisit this. My current approach works only if `max_per_node == 1`. If `max_per_node == 2`, for example, then we shouldn't mark a node unavailable until it has two reserved instances.

I'll have to think about how to handle that... We could use a hash table to track reserved instances per node in addition to the `reserved` counter variable. Maybe repurpose the `pe_node_t:count` member if it doesn't conflict with an existing use. The parent's `count` member already tracks the total instances assigned to a node. I'd rather not mess with that during these scratch assignments, if I can avoid it. I guess worst case I could make a copy of one of the allowed_nodes tables.

Reply:
Updated commit to add a `reserved_table` to `assign_instances_early()`. If I didn't typo anything, that should take care of checking against `max_per_node`. I'd be a little bit surprised if any of our regression tests exercise that aspect.

`assign_instances_early()` is longer than I would like it to be, but it's mostly comments and debug messages. I'm not sure it's worth functionizing any of it. Maybe the management of `reserved_table`, if anything.

I'm going to bed.

Reply:
Gonna see if I can rearrange things so that `ban_unavailable_allowed_nodes` takes care of all this for me

Reply:
I did simplify the `assign_instances_early()` function so that it lets `ban_unavailable_allowed_nodes()` do the work, but the colocation issues are still blocking me.